### PR TITLE
support for ignore unknown ids in time series

### PR DESCRIPTION
**Pull request by sanderland**

### CODE REVIEW

File: cognite/client/_api/datapoints.py [hunk 1]
```diff
@@ -84,7 +86,7 @@ def retrieve(
                 ...                         external_id={"externalId": "1", "aggregates": ["max"]},
                 ...                         start="1d-ago", end="now", granularity="1h")
         """
-        fetcher = DatapointsFetcher(client=self)
+        fetcher = DatapointsFetcher(client=self, ignore_unknown_ids=ignore_unknown_ids)
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]: To stay aligned with how we handle all other parameters, it should rather be an argument to `DatapointsQuery`.

File: cognite/client/_api/datapoints.py [hunk 2]
```diff
@@ -99,7 +101,7 @@ def retrieve(
             limit=limit,
         )
         dps_list = fetcher.fetch(query)
-        if is_single_id:
+        if is_single_id and not ignore_unknown_ids:
             return dps_list[0]
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]:
```
if is_single_id:
    if ignore_unknown_ids is True:
        return None
    return dps_list[0]
```
And change the return type hint to `Union[Datapoints, DatapointsList, None]`

File: cognite/client/_api/datapoints.py [hunk 3]
```diff
@@ -165,7 +172,7 @@ def retrieve_latest(
         return DatapointsList._load(res, cognite_client=self._cognite_client)
 
     def query(
-        self, query: Union[DatapointsQuery, List[DatapointsQuery]]
+        self, query: Union[DatapointsQuery, List[DatapointsQuery]], ignore_unknown_ids=False
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]: Should be on the `DatapointsQuery` instead.

File: cognite/client/_api/datapoints.py (hunk 2 above)
Comment (sanderland) [reply cognite/client/_api/datapoints.py]: having two return types for retrieve is already a pain, do you really think three different ones will help?

File: cognite/client/_api/datapoints.py (hunk 3 above)
Comment (sanderland) [reply cognite/client/_api/datapoints.py]: I think it is prefered by users as is, but we can ask them

File: cognite/client/_api/datapoints.py (hunk 2 above)
Comment (sanderland) [reply cognite/client/_api/datapoints.py]: to follow up:
2 requested, ignoreunknown False -> DatapointsList of length 2
single requested, ignoreunknown False -> Datapoints
2 requested, 1 found, ignoreunknown True -> DatapointsList of length 1
2 requested, 0 found, ignoreunknown True -> DatapointsList of length 0
single requested, 0 found ignoreunknown True -> new behaviour: None ?
agreed?

File: cognite/client/_api/datapoints.py (hunk 1 above)
Comment (sanderland) [reply cognite/client/_api/datapoints.py]: done

File: cognite/client/_api/datapoints.py (hunk 2 above)
Comment (erlendvollset) [reply cognite/client/_api/datapoints.py]: The nested if there should of course be `if ignore_unknown_ids and len(dps_list) == 0`. Sorry..

IMO it is better to have behaviour consistent with other methods for single retrieval than returning an empty list upon single retrieval not found.

Comment (sanderland) [reply cognite/client/_api/datapoints.py]: I fixed it in the last commit. It's a bit painful in retrieve_dataframe to handle 3 return types, but I think it's ok now

File: cognite/client/_api/datapoints.py [hunk 4]
```diff
@@ -97,9 +99,12 @@ def retrieve(
             granularity=granularity,
             include_outside_points=include_outside_points,
             limit=limit,
+            ignore_unknown_ids=ignore_unknown_ids,
         )
         dps_list = fetcher.fetch(query)
         if is_single_id:
+            if len(dps_list) == 0 and ignore_unknown_ids is True:
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]: Return type hint and and return value in docstring are still wrong.

File: cognite/client/_api/datapoints.py [hunk 5]
```diff
@@ -389,6 +399,7 @@ def retrieve_dataframe(
         limit: int = None,
         include_aggregate_name=True,
         complete: str = None,
+        ignore_unknown_ids=False,
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]: Type hint

File: cognite/client/_api/datapoints.py [hunk 6]
```diff
@@ -528,6 +551,7 @@ def retrieve_dataframe_dict(
             str, List[str], Dict[str, Union[int, List[str]]], List[Dict[str, Union[int, List[str]]]]
         ] = None,
         limit: int = None,
+        ignore_unknown_ids=False,
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py]: Type hint

**Pull request merged**